File: leetcode/56/main.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) <= 1:
            return intervals

        intervals.sort()
        current_start: int = intervals[0][0]
        current_end: int = intervals[0][1]
        output: list[list[int,int]] = []
        for (interval_start, interval_end) in intervals:
            if interval_start <= current_end:
                if interval_end > current_end:
                    current_end = interval_end
            else:
                output.append([current_start, current_end])
                current_start = interval_start
                current_end = interval_end
        output.append([current_start, current_end])
        return output

File: leetcode/56/main.go
/* Notes:
    - it would probably help to sort all intervals by start
    - for each interval, search all other intervals for overlapping ones. You will only need to search intervals with start less than or equal to current intervals end


*/

import (
    "fmt"
)

type intervalsSorted [][]int

func (a intervalsSorted) Len() int { return len(a) }
func (a intervalsSorted) Swap(i, j int) { a[i], a[j] = a[j], a[i]}
func (a intervalsSorted) Less(i, j int) bool { return a[i][0] < a[j][0]}

func merge(intervals [][]int) [][]int {

    // sort all intervals by start
    sort.Sort(intervalsSorted(intervals))

    // search for overlapping intervals
    for i:=0; i<len(intervals)-1; i++ {
        current := intervals[i]
        compare := intervals[i+1]
        if compare[0] > current[1] {
            continue
        } else {
            // merge
            newInterval := []int{current[0], compare[1]}
            if current[0] > compare[0] {
                newInterval[0] = compare[0]
            }
            if current[1] > compare[1] {
                newInterval[1] = current[1]
            }
            // update intervals
            if i == len(intervals)-2 {
                intervals = append(intervals[:i], newInterval)
            } else {
                intervals = append(append(intervals[:i], newInterval), intervals[i+2:]...)
            }
            i--
        }        

    }

    return intervals
}
